In [2]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

import numpy as np # library to handle data in a vectorized manner

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
#=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: | 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/linux-64::anaconda==5.3.1=py37_0
  - defaults/linux-64::astropy==3.0.4=py37h14c3975_0
  - defaults/linux-64::bkcharts==0.2=py37_0
  - defaults/linux-64::blaze==0.11.3=py37_0
  - defaults/linux-64::bokeh==0.13.0=py37_0
  - defaults/linux-64::bottleneck==1.2.1=py37h035aef0_1
  - defaults/linux-64::dask==0.19.1=py37_0
  - defaults/linux-64::datashape==0.5.4=py37_1
  - defaults/linux-64::mkl-service==1.1.2=py37h90e4bf4_5
  - defaults/linux-64::numba==0.39.0=py37h04863e7_0
  - defaults/linux-64::numexpr==2.6.8=py37hd89afb7_0
  - defaults/linux-64::odo==0.5.1=py37_0
  - defaults/linux-64::pytables==3.4.4=py37ha205bf6_0
  - defaults/linux-64::pytest-arraydiff==0.2=py37h39e3cac_0
  - defaults/linux-64::pytest-astropy==0.4.0=py37_0
  - defaults/linux-64::pytest-doctestplus==0.1.3=py37_0
  - defaults/linux-64::pywavelets==1.0.0=py37h

In [3]:
website_url = requests.get("https://es.wikipedia.org/wiki/Anexo:Localidades_de_Bogot%C3%A1").text

In [4]:
soup = BeautifulSoup(website_url,'lxml')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="es">
 <head>
  <meta charset="utf-8"/>
  <title>
   Anexo:Localidades de Bogotá - Wikipedia, la enciclopedia libre
  </title>
  <script>
   document.documentElement.className = document.documentElement.className.replace( /(^|\s)client-nojs(\s|$)/, "$1client-js$2" );
  </script>
  <script>
   (window.RLQ=window.RLQ||[]).push(function(){mw.config.set({"wgCanonicalNamespace":"Anexo","wgCanonicalSpecialPageName":false,"wgNamespaceNumber":104,"wgPageName":"Anexo:Localidades_de_Bogotá","wgTitle":"Localidades de Bogotá","wgCurRevisionId":114573108,"wgRevisionId":114573108,"wgArticleId":6319494,"wgIsArticle":true,"wgIsRedirect":false,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Localidades de Bogotá","Anexos:Bogotá"],"wgBreakFrames":false,"wgPageContentLanguage":"es","wgPageContentModel":"wikitext","wgSeparatorTransformTable":[",\t."," \t,"],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMont

In [6]:
My_table = soup.find('table',{'class':'sortable wikitable'})
My_table

<table border="1" class="sortable wikitable">
<tbody><tr bgcolor="silver">
<th width="23"><center>Nº
</center></th>
<th width="110"><center> Localidad </center>
</th>
<th width="90"><center> Códigos Postales </center>
</th>
<th><center> Superficie km²<sup class="reference separada" id="cite_ref-2"><a href="#cite_note-2"><span class="corchete-llamada">[</span>2<span class="corchete-llamada">]</span></a></sup>​</center>
</th>
<th><center> Población<sup class="reference separada" id="cite_ref-3"><a href="#cite_note-3"><span class="corchete-llamada">[</span>3<span class="corchete-llamada">]</span></a></sup>​</center>
</th>
<th><center> Densidad hab/km²</center>
</th></tr>
<tr>
<td><b>01</b>
</td>
<td><a href="/wiki/Usaqu%C3%A9n" title="Usaquén">Usaquén</a>
</td>
<td>11<b>01</b>11-11<b>01</b>51
</td>
<td>65.31
</td>
<td>501 999
</td>
<td>7 686.4
</td></tr>
<tr>
<td><b>02</b>
</td>
<td><a href="/wiki/Chapinero" title="Chapinero">Chapinero</a>
</td>
<td>11<b>02</b>11-11<b>02</b>31
</td>
<td>3

In [7]:
links = My_table.findAll('a')
links

[<a href="#cite_note-2"><span class="corchete-llamada">[</span>2<span class="corchete-llamada">]</span></a>,
 <a href="#cite_note-3"><span class="corchete-llamada">[</span>3<span class="corchete-llamada">]</span></a>,
 <a href="/wiki/Usaqu%C3%A9n" title="Usaquén">Usaquén</a>,
 <a href="/wiki/Chapinero" title="Chapinero">Chapinero</a>,
 <a href="/wiki/Santa_Fe_(Bogot%C3%A1)" title="Santa Fe (Bogotá)">Santa Fe</a>,
 <a href="/wiki/San_Crist%C3%B3bal_(Bogot%C3%A1)" title="San Cristóbal (Bogotá)">San Cristóbal</a>,
 <a href="/wiki/Usme" title="Usme">Usme</a>,
 <a href="/wiki/Tunjuelito" title="Tunjuelito">Tunjuelito</a>,
 <a href="/wiki/Bosa_(Bogot%C3%A1)" title="Bosa (Bogotá)">Bosa</a>,
 <a href="/wiki/Kennedy_(Bogot%C3%A1)" title="Kennedy (Bogotá)">Kennedy</a>,
 <a href="/wiki/Fontib%C3%B3n" title="Fontibón">Fontibón</a>,
 <a href="/wiki/Engativ%C3%A1" title="Engativá">Engativá</a>,
 <a href="/wiki/Suba" title="Suba">Suba</a>,
 <a href="/wiki/Barrios_Unidos_(Bogot%C3%A1)" title="Barrios 

In [8]:
Localidades = []
for link in links:
    Localidades.append(link.get('title'))
    
print(Localidades)

del Localidades[0:2]

Localidades

[None, None, 'Usaquén', 'Chapinero', 'Santa Fe (Bogotá)', 'San Cristóbal (Bogotá)', 'Usme', 'Tunjuelito', 'Bosa (Bogotá)', 'Kennedy (Bogotá)', 'Fontibón', 'Engativá', 'Suba', 'Barrios Unidos (Bogotá)', 'Teusaquillo', 'Los Mártires', 'Antonio Nariño (Bogotá)', 'Puente Aranda', 'La Candelaria', 'Rafael Uribe Uribe (Bogotá)', 'Ciudad Bolívar (Bogotá)', 'Sumapaz (Bogotá)']


['Usaquén',
 'Chapinero',
 'Santa Fe (Bogotá)',
 'San Cristóbal (Bogotá)',
 'Usme',
 'Tunjuelito',
 'Bosa (Bogotá)',
 'Kennedy (Bogotá)',
 'Fontibón',
 'Engativá',
 'Suba',
 'Barrios Unidos (Bogotá)',
 'Teusaquillo',
 'Los Mártires',
 'Antonio Nariño (Bogotá)',
 'Puente Aranda',
 'La Candelaria',
 'Rafael Uribe Uribe (Bogotá)',
 'Ciudad Bolívar (Bogotá)',
 'Sumapaz (Bogotá)']

In [9]:
df = pd.DataFrame()
df['Localidades'] = Localidades
df

Localidades
0                       Usaquén
1                     Chapinero
2             Santa Fe (Bogotá)
3        San Cristóbal (Bogotá)
4                          Usme
5                    Tunjuelito
6                 Bosa (Bogotá)
7              Kennedy (Bogotá)
8                      Fontibón
9                      Engativá
10                         Suba
11      Barrios Unidos (Bogotá)
12                  Teusaquillo
13                 Los Mártires
14      Antonio Nariño (Bogotá)
15                Puente Aranda
16                La Candelaria
17  Rafael Uribe Uribe (Bogotá)
18      Ciudad Bolívar (Bogotá)
19             Sumapaz (Bogotá)

In [10]:
def get_coords_local(localidad, output_as='center'):
    """
    get the bounding box of a locality in WGS84 given its name

    Parameters
    ----------
    localidad : str
        name of the country in english and lowercase
    output_as : 'str
        chose from 'boundingbox' or 'center'. 
         - 'boundingbox' for [latmin, latmax, lonmin, lonmax]
         - 'center' for [latcenter, loncenter]

    Returns
    -------
    output : list
        list with coordinates as str
    """
    # create url
    url = '{0}{1}{2}'.format('http://nominatim.openstreetmap.org/search.php?q=',
                             localidad+', Bogota, Bogota Capital District',
                             '&format=json&polygon=0')
    response = requests.get(url).json()[0]

    # parse response to list
    if output_as == 'boundingbox':
        lst = response[output_as]
        output = [float(i) for i in lst]
    if output_as == 'center':
        lst = [response.get(key) for key in ['lat','lon']]
        output = [float(i) for i in lst]
    return output

In [11]:
df2 = df.copy()

latitudeCln = []
longitudeCln = []
for index, row in df2.iterrows():
    print(row[0])
    lat, long = get_coords_local(localidad=row[0], output_as='center')
    latitudeCln.append(lat)
    longitudeCln.append(long)

df2['Latitude'] = latitudeCln
df2['Longitude'] = longitudeCln

df2.shape

Usaquén
Chapinero
Santa Fe (Bogotá)
San Cristóbal (Bogotá)
Usme
Tunjuelito
Bosa (Bogotá)
Kennedy (Bogotá)
Fontibón
Engativá
Suba
Barrios Unidos (Bogotá)
Teusaquillo
Los Mártires
Antonio Nariño (Bogotá)
Puente Aranda
La Candelaria
Rafael Uribe Uribe (Bogotá)
Ciudad Bolívar (Bogotá)
Sumapaz (Bogotá)


(20, 3)

In [12]:
df2

Localidades  Latitude  Longitude
0                       Usaquén  4.695047 -74.031493
1                     Chapinero  4.647120 -74.063458
2             Santa Fe (Bogotá)  4.593766 -74.034314
3        San Cristóbal (Bogotá)  4.548658 -74.047473
4                          Usme  4.411136 -74.129108
5                    Tunjuelito  4.560148 -74.128922
6                 Bosa (Bogotá)  4.625492 -74.200280
7              Kennedy (Bogotá)  4.629682 -74.149935
8                      Fontibón  4.678737 -74.146988
9                      Engativá  4.696628 -74.106120
10                         Suba  4.761197 -74.082518
11      Barrios Unidos (Bogotá)  4.669679 -74.075483
12                  Teusaquillo  4.642343 -74.087217
13                 Los Mártires  4.608375 -74.086538
14      Antonio Nariño (Bogotá)  4.588253 -74.097455
15                Puente Aranda  4.619234 -74.106763
16                La Candelaria  4.596515 -74.073492
17  Rafael Uribe Uribe (Bogotá)  4.573490 -74.119208
18      Ciudad Bolívar (Bogotá)  4.492632 -74.143119
19             Sumapaz (Bogotá)  4.097379 -74.342307

In [13]:
address = 'Bogotá, Colombia'

geolocator = Nominatim(user_agent="capstoneProject")
location = geolocator.geocode(address, timeout=60, exactly_one=True)
latitude = location.latitude
longitude = location.longitude
print('The decimal coordinates of Bogotá are {}, {}.'.format(latitude, longitude))

The decimal coordinates of Bogotá are 4.5986988, -74.0758477.


In [14]:
# make a copy of the dataframe to get it simply back if needed
df3 = df2.copy()
df3

Localidades  Latitude  Longitude
0                       Usaquén  4.695047 -74.031493
1                     Chapinero  4.647120 -74.063458
2             Santa Fe (Bogotá)  4.593766 -74.034314
3        San Cristóbal (Bogotá)  4.548658 -74.047473
4                          Usme  4.411136 -74.129108
5                    Tunjuelito  4.560148 -74.128922
6                 Bosa (Bogotá)  4.625492 -74.200280
7              Kennedy (Bogotá)  4.629682 -74.149935
8                      Fontibón  4.678737 -74.146988
9                      Engativá  4.696628 -74.106120
10                         Suba  4.761197 -74.082518
11      Barrios Unidos (Bogotá)  4.669679 -74.075483
12                  Teusaquillo  4.642343 -74.087217
13                 Los Mártires  4.608375 -74.086538
14      Antonio Nariño (Bogotá)  4.588253 -74.097455
15                Puente Aranda  4.619234 -74.106763
16                La Candelaria  4.596515 -74.073492
17  Rafael Uribe Uribe (Bogotá)  4.573490 -74.119208
18      Ciudad Bolívar (Bogotá)  4.492632 -74.143119
19             Sumapaz (Bogotá)  4.097379 -74.342307

In [15]:
# create map of Bogotá using latitude and longitude values
map_bogota = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, local in zip(df3['Latitude'], df3['Longitude'], df3['Localidades']):
    label = '{}'.format(local)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_bogota)  
    
map_bogota

In [16]:
# function to repeat the exploring process to all the neighborhoods in Toronto
import urllib
def getNearbyVenues(names, latitudes, longitudes, radius=5000, categoryIds=''):
    try:
        venues_list=[]
        for name, lat, lng in zip(names, latitudes, longitudes):
            #print(name)

            # create the API request URL
            url = 'https://api.foursquare.com/v2/venues/search?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION, lat, lng, radius, LIMIT)

            if (categoryIds != ''):
                url = url + '&categoryId={}'
                url = url.format(categoryIds)

            # make the GET request
            response = requests.get(url).json()
            results = response["response"]['venues']

            # return only relevant information for each nearby venue
            for v in results:
                success = False
                try:
                    category = v['categories'][0]['name']
                    success = True
                except:
                    pass

                if success:
                    venues_list.append([(
                        name, 
                        lat, 
                        lng, 
                        v['name'], 
                        v['location']['lat'], 
                        v['location']['lng'],
                        v['categories'][0]['name']
                    )])

        nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
        nearby_venues.columns = ['Localidad', 
                  'Localidad Latitude', 
                  'Localidad Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    except:
        print(url)
        print(response)
        print(results)
        print(nearby_venues)

    return(nearby_venues)

In [23]:
limit = 500 # limit of number of venues returned by Foursquare API
radius = 5000 # define radius
CLIENT_ID = 'FZYS253SK224TRMBI31QKSDYVKGCFVI1X4YRXD2BQYIB1ALC'
CLIENT_SECRET = 'X4QJ13ESLW2ZKK1OYGPPD0KRWOE2ECRELXMHK31VCWN4H2Z2'
VERSION = '20181020'

In [25]:
# function to add markers for given venues to map
def addToMap(df, color, existingMap):
    for lat, lng, local, venue, venueCat in zip(df['Venue Latitude'], df['Venue Longitude'], df['Localidad'], df['Venue'], df['Venue Category']):
        label = '{} ({}) - {}'.format(venue, venueCat, local)
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            popup=label,
            color=color,
            fill=True,
            fill_color=color,
            fill_opacity=0.7).add_to(existingMap)

	Localidad	Latitude	Longitude	Burger	High Schools	Universities	Offices
0	Usaquén	4.695047	-74.031493	31.0	1.0	3.0	50.0
1	Chapinero	4.649917	-74.046269	6.0	0.0	2.0	50.0
2	Santa Fe (Bogotá)	4.593766	-74.034314	0.0	0.0	0.0	0.0
3	San Cristóbal (Bogotá)	4.548658	-74.047473	0.0	0.0	0.0	0.0
4	Usme	4.411136	-74.129108	0.0	0.0	0.0	0.0
5	Tunjuelito	4.560148	-74.128922	4.0	2.0	0.0	50.0
6	Bosa (Bogotá)	4.625515	-74.200328	0.0	1.0	0.0	24.0
7	Kennedy (Bogotá)	4.629451	-74.149927	27.0	1.0	1.0	50.0
8	Fontibón	4.678737	-74.146988	7.0	3.0	2.0	50.0
9	Engativá	4.696628	-74.106120	11.0	4.0	1.0	50.0
10	Suba	4.761197	-74.082518	2.0	4.0	0.0	26.0
11	Barrios Unidos (Bogotá)	4.669679	-74.075483	17.0	3.0	1.0	50.0
12	Teusaquillo	4.641244	-74.086336	18.0	0.0	8.0	50.0
13	Los Mártires	4.608375	-74.086538	12.0	1.0	3.0	50.0
14	Antonio Nariño (Bogotá)	4.588253	-74.097455	9.0	4.0	4.0	50.0
15	Puente Aranda	4.619234	-74.106763	11.0	0.0	2.0	50.0
16	La Candelaria	4.596515	-74.073492	48.0	2.0	37.0	50.0
17	Rafael Uribe Uribe (Bogotá)	4.573490	-74.119208	7.0	5.0	0.0	50.0
18	Ciudad Bolívar (Bogotá)	4.492632	-74.143119	0.0	0.0	0.0	0.0
19	Sumapaz (Bogotá)	4.098800	-74.341530	0.0	0.0	0.0	0.0

	Localidad	Score
16	La Candelaria	109.5
14	Antonio Nariño (Bogotá)	101.0
8	Fontibón	99.0
17	Rafael Uribe Uribe (Bogotá)	98.0
5	Tunjuelito	98.0
1	Chapinero	97.0
9	Engativá	94.5
12	Teusaquillo	94.0
13	Los Mártires	93.5
15	Puente Aranda	92.0
11	Barrios Unidos (Bogotá)	87.5
7	Kennedy (Bogotá)	75.5
0	Usaquén	74.5
10	Suba	54.0
6	Bosa (Bogotá)	49.0
4	Usme	0.0
3	San Cristóbal (Bogotá)	0.0
2	Santa Fe (Bogotá)	0.0
18	Ciudad Bolívar (Bogotá)	0.0
19	Sumapaz (Bogotá)	0.0